docs/Tutorial_DiscordMERLIN.ipynb

```diff
@@ -0,0 +1,1401 @@
+{
```

Comment:
<p>I was thinking about this and, for the non-normalized section, can you please add a small note that, in reality, <code>Tmax</code>  and <code>Tmin</code>  should realy be:</p><p> </p><pre class="ql-syntax" spellcheck="false">Tmax = np.max(T[np.isfinite(T)])
Tmin = np.min(T[np.isfinite(T)])
</pre><p> </p><p>This way, if a time series contains <code>np.inf</code> , <code>-np.inf</code> , or <code>np.nan</code>  then we avoid <code>Tmax = np.inf</code>  and <code>Tmin = -np.inf</code> , which would make <code>min_dist</code>  useless.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/TDAmeritrade/stumpy/pull/505/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Reply:
I will add a note regarding this matter. By the way, can `_mass` (or `_mass_absolute`) consider time stamps with `np.inf` values while calculating the sliding distance between a Q and a time series T?

Reply:
> I will add a note regarding this matter. So, your comment is just for the sake of explanation as I am expecting such scenarios are already considered in the STUMPY's pre-processing stage.

Thank you. Yes, I just didn't want to forget this when we implemented `min_dist` here and in other non-nomrmalized functions like `stumpy.stimp`

> By the way, can _mass (or _mass_absolute) consider time stamps with np.inf values while calculating the sliding distance between a Q and a time series T? I mean, if there is a infinite value, should we bother detecting that as discord? Because I do not think z-normalization work on a subsequence that has np.inf. And, those discords can be visually detected by just plotting the time series.

So, both `core.preprocess` and `core.preprocess_non_normalized` both make a copy of your data and essentially captures and converts all non-finite numbers to zero `T[~np.isfinite(T)] = 0.0` before returning the time series. Thus, the "information" that an element/or subsequence contains a non-finite number is implicitly captured in `M_T` (i.e., the mean of `T`) and `T_subseq_isfinite` for z-norm and non-norm, respectively.

```
import numpy as np
from stumpy import core

T = np.random.rand(10)
T[0] = np.inf
core.preprocess(T, 5)

(array([0.        , 0.80747123, 0.05026668, 0.60174256, 0.71609755,  # `T` with `np.inf`/`np.nan` converted to "0"
        0.30642316, 0.57479348, 0.89419958, 0.5403805 , 0.52095175]),  
 array([       inf, 0.49640024, 0.44986469, 0.61865127, 0.60637885,  # Notice the first element of `M_T` is `np.inf`
        0.56734969]),
 array([0.29416892, 0.27973399, 0.24075218, 0.19241541, 0.19504143, 
        0.18859958]))
```

So, the way that we've abstracted the functions, `_mass` and `_mass_absolute` ALWAYS assumes that the input data does not contain `np.inf`/`np.nan`. This is by design and these two functions are "private" and assumes that you know what you are doing and it is up to the user to clean things up after the fact. Most users should NEVER use `_mass` or `_mass_absolute` directly. Instead, they should use `core.mass` instead which is a wrapper around both of these functions and performs the clean up for you. However, they can be slower.

Thus, when you use `core._mass` or `core._mass_absolute`, I strongly recommend looking `core.mass` and `core.mass_absolute` and make sure that you are performing the same checking/cleanup steps before/after you call those functions.

For example, in `core.mass_absolute`, you'll see that it starts with many, many checks of the input data and, at the end, it does

```
distance_profile[:] = _mass_absolute(Q_squared, T_squared, QT)
distance_profile[~T_subseq_isfinite] = np.inf
```

Let me know if that makes sense.

Reply:
Thanks for your nice explanation.

I thought `core.mass` does not accept sliding mu / std of T. I don't know why I overlooked such an obvious thing! I took a look at the function again and it is just there!! I used `core.mass` without using the pre-calculated sliding mu/std, and I noticed it was slow. That is the reason I decided to use `core._mass` instead.

So, I can switch back to using `core.mass`. Its disadvantage is that I still need to re-calculate mu/std of Q(query with size `m`) in each iteration and check/cleanup things. Its advantage is that it does the pre-/post- processing task itself.

Since Q(query) is coming from the same time series, I think we can do a check on `T` first and then just use `_mass`. I will think about. Do you have any particular suggestion on this?  I can do the changes in the next commits after reviewing the recent push.

Reply:
I agree. For something like Merlin that is meant to be a new public function, we would want to put in the work to make the backend as solid and as possible. Otherwise, nobody will ever improve that code. :)

To be unambiguous, please use `core._mass` and please perform the pre- and post-processing. Everybody will benefit.